modules/sitemaps/sitemaps.php

```diff
@@ -0,0 +1,136 @@
+<?php
+/**
+ * @package Polylang
+ */
+
+/**
+ * Handles the core sitemaps.
+ *
+ * @since 2.8
+ */
+class PLL_Sitemaps {
+	/**
+	 * A reference to teh PLL_Model instance.
+	 *
+	 * @since 2.8
+	 *
+	 * @var PLL_Model
+	 */
+	protected $model;
+
+	/**
+	 * A reference to the current language.
+	 *
+	 * @since 2.8
+	 *
+	 * @var PLL_Language
+	 */
+	protected $curlang;
+
+	/**
+	 * Constructor.
+	 *
+	 * @since 2.8
+	 *
+	 * @param object $polylang Main Polylang object.
+	 */
+	public function __construct( $polylang ) {
+		$this->model = &$polylang->model;
+		$this->curlang = &$polylang->curlang;
+	}
+
+	/**
+	 * Setups actions and filters.
+	 *
+	 * @since 2.8
+	 */
+	public function init() {
+		add_action( 'template_redirect', array( $this, 'set_language' ), 5 ); // Before WP_Sitemaps.
+		add_filter( 'pll_home_url_white_list', array( $this, 'home_url_white_list' ) );
+		add_filter( 'rewrite_rules_array', array( $this, 'rewrite_rules' ) );
+		add_filter( 'wp_sitemaps_register_providers', array( $this, 'providers' ), 99 ); // 99 in an attempt to have all sitemaps providers in our filter.
+	}
+
+	/**
+	 * Sets the language for the current sitemap being rendered.
+	 *
+	 * @since 2.8
+	 */
+	public function set_language() {
+		$qv = get_query_var( 'sitemap' );
+		if ( $qv ) {
+			$arr = explode( '-', $qv );
+			$lang = end( $arr );
+			$this->curlang = $this->model->get_language( $lang );
+		}
+	}
+
+	/**
+	 * Whitelists the home url filter for the sitemaps
+	 *
+	 * @since 2.8
+	 *
+	 * @param array $whitelist White list.
+	 * @return array;
+	 */
+	public function home_url_white_list( $whitelist ) {
+		$whitelist[] = array( 'file' => 'sitemaps' );
```

Comment:
Ain't that too generic for the usage we are doing it here? Would we want to automatically whitelist any plugin that have 'sitemap' in its files names?

I believe you have made it that way to not have to change it every time the sitemap feature changes its code, but would it cover all the cases? What happens the call comes from a function in the `functions.php` file for example? What if a third party plugin triggers a call from a wp-sitemap class, but from the context of this 3rd party plugins files, thus not bearing the 'sitemap' mention in its file name?

Reply:
See above. For 3rd party plugins, we need to whitelist the filter on a case by case basis. As done for Yoast: https://github.com/polylang/polylang/blob/a73de2b21fbfde0587a7c3df71f309cebf3e6a73/integrations/wpseo/wpseo.php#L40

Reply:
What I'm trying to tell, is that the [PLL_Frontend_Filter::home_url()](https://github.com/polylang/polylang/blob/7bea3700a7293f8ef31b700fa0fe84647420c53b/frontend/frontend-filters-links.php#L308) method will whitelist calls from files where only a portion of the filename matches the string.

So whitelisting the `sitemaps` string seems to imply that every call to the home url coming from a file having `sitemap` in its filename will come from WordPress Core's sitemap feature, and every call to the home url from WordPress Core's sitemap feature will come from a file with `sitemap` in it's filename. I don't believe that this will always be the case, and I came with the examples listed above.